File: src/main/java/org/normal/api/java/context/Context.java
package org.normal.api.java.context;

import java.util.Optional;
import java.util.UUID;

public interface Context<T extends ContextType> {

    T getType();

    <Z> Optional<Z> getData(String key, Class<Z> type);

    default Optional<String> getId() {
        return getData("id", String.class);
    }

    default Optional<String> getName() {
        return getData("name", String.class);
    }

    default Optional<UUID> getUuid() {
        return getData("uuid", UUID.class);
    }

    <Z> void putData(String key, Z value);

}

File: src/main/java/org/normal/api/kt/context/Context.kt
package org.normal.api.kt.context

import java.util.*

interface Context {

    val type: ContextType

    fun <data : Any> getData(key: String, type: Class<data>): data?

    fun <data : Any> putData(key: String, value: data)

    fun getId(): String? = getData("id", String::class.java)

    fun getName(): String? = getData("name", String::class.java)

    fun getUuid(): UUID? = getData("uuid", UUID::class.java)

}